libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp

```diff
@@ -0,0 +1,251 @@
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#if CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BEBOP
+#include "OpticalFlow_Onboard.h"
+
+#include <stdio.h>
+#include <pthread.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <linux/v4l2-mediabus.h>
+
+#include "CameraSensor_Mt9v117.h"
+#include "GPIO.h"
+#include "PWM_Sysfs.h"
+
+#define OPTICAL_FLOW_ONBOARD_RTPRIO 11
+
+extern const AP_HAL::HAL& hal;
+
+using namespace Linux;
+
+void OpticalFlow_Onboard::init(AP_HAL::OpticalFlow::Gyro_Cb get_gyro)
+{
+    uint32_t top, left;
+    uint32_t crop_width, crop_height;
+    uint32_t memtype = V4L2_MEMORY_MMAP;
+    unsigned int nbufs = 0;
+    int ret;
+    pthread_attr_t attr;
+    struct sched_param param = {
+        .sched_priority = OPTICAL_FLOW_ONBOARD_RTPRIO
+    };
+
+    if (_initialized) {
+        return;
+    }
+
+    _get_gyro = get_gyro;
+
+    _videoin = new VideoIn;
+    _pwm = new PWM_Sysfs_Bebop(BEBOP_CAMV_PWM);
+    _pwm->set_freq(BEBOP_CAMV_PWM_FREQ);
+    _pwm->enable(true);
+
+    _camerasensor = new CameraSensor_Mt9v117(HAL_OPTFLOW_ONBOARD_SUBDEV_PATH,
+                                             hal.i2c, 0x5D, MT9V117_QVGA,
+                                             BEBOP_GPIO_CAMV_NRST,
+                                             BEBOP_CAMV_PWM_FREQ);
+    if (!_camerasensor->set_format(HAL_OPTFLOW_ONBOARD_SENSOR_WIDTH,
+                                   HAL_OPTFLOW_ONBOARD_SENSOR_HEIGHT,
+                                   V4L2_MBUS_FMT_UYVY8_2X8)) {
+        AP_HAL::panic("OpticalFlow_Onboard: couldn't set subdev fmt\n");
+    }
+    const char* device_path = HAL_OPTFLOW_ONBOARD_VDEV_PATH;
+    memtype = V4L2_MEMORY_MMAP;
+    nbufs = HAL_OPTFLOW_ONBOARD_NBUFS;
+    _width = HAL_OPTFLOW_ONBOARD_OUTPUT_WIDTH;
+    _height = HAL_OPTFLOW_ONBOARD_OUTPUT_HEIGHT;
+    crop_width = HAL_OPTFLOW_ONBOARD_CROP_WIDTH;
+    crop_height = HAL_OPTFLOW_ONBOARD_CROP_HEIGHT;
+    top = 0;
+    /* make the image square by cropping in the center */
+    left = (HAL_OPTFLOW_ONBOARD_OUTPUT_WIDTH -
+            HAL_OPTFLOW_ONBOARD_OUTPUT_HEIGHT) / 2;
+    _format = V4L2_PIX_FMT_NV12;
+
+    if (device_path == NULL ||
+        !_videoin->open_device(device_path, memtype)) {
+        AP_HAL::panic("OpticalFlow_Onboard: couldn't open "
+                      "video device");
+    }
+
+    if (!_videoin->set_format(&_width, &_height, &_format, &_bytesperline,
+                              &_sizeimage)) {
+        AP_HAL::panic("OpticalFlow_Onboard: couldn't set video format");
+        return;
+    }
+
+    if (_format != V4L2_PIX_FMT_NV12 && _format != V4L2_PIX_FMT_GREY) {
+        AP_HAL::panic("OpticalFlow_Onboard: planar or monochrome format needed");
+    }
+
+    if (!_videoin->set_crop(left, top, crop_width, crop_height)) {
```

Comment:
Ricardo,
I suggest we continue this discussion on google groups if you don't mind. I don't want to give the impression that this code is under development because it is functional on the Bebop.
We'll then sort out how to make it work on your board.